src/main/java/com/ibm/crypto/plus/provider/ock/Digest.java

```diff
@@ -340,15 +343,28 @@ protected static boolean validId(long id) {
         return (id != 0L);
     }
 
-    @Override
-    synchronized public Object clone() throws CloneNotSupportedException {
-        Digest copy = (Digest) super.clone();
-        
+    /**
+     * Clones a given Digest.
+     */
+    public synchronized Object clone() throws CloneNotSupportedException {
+        // Create new Digest instance and copy all relevant fields into the copy.
+        // Clones do not make use of the cache so always set the value of
+        // contextFromQueue to false to ensure that the context is later freed
+        // correctly.
+        Digest copy = new Digest();
+        copy.digestLength = this.digestLength;
+        copy.algIndx = this.algIndx;
+        copy.digestAlgo = new String(this.digestAlgo);
+        copy.needsReinit = this.needsReinit;
+        copy.ockContext = this.ockContext;
         copy.contextFromQueue = false;
+
+        // Allocate a new context for the digestId and copy all state information from our
+        // original context into the copy. 
         try {
             copy.digestId = NativeInterface.DIGEST_copy(
                 this.ockContext.getId(), getId());
-            if (copy.digestId == 0) {
+            if (0 == copy.digestId) {
```

Comment:
We use this convention of putting the constant first in the C code. Is that something we want to do in Java code as well?

Reply:
Yes good question. In general I do think its a good idea for us to typically put the constant first. The main argument I have heard over time is that the compile will prevent mistakes such as forgetting the `==` and instead putting a `=` into the if statement. Using a constant ensures that the compile fails if we were to do `if (0 = copy.digestId)` by mistake. Same thinking also applies to the Java code. Id say we should fix these as we go along with changes.